### PR TITLE
[Attempt 3] Implement FileManager.url{s,}(for:…) and NSSearchPathForDirectoriesInDomains.

### PR DESCRIPTION
**Pull request by millenomi**
This re-reproposes the patch at #1597 that was reverted by #1598. It should fix the issues from the previous two attempts. (Whoops.)

### CODE REVIEW

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
cc @rudkx @shahmishal I'm attempting again. I'm trying to figure out why the buildbot found that issue but both local Ubuntu 16.10 and Mac builds didn't find this.

**Comment by felix91gr** [comment]
> Ubuntu 16.10

Did you try it with 16.04? Since it's LTS, it's usually more representative of how Ubuntu behaves :)

**Comment by rudkx** [comment]
We can certainly try merging again and see what the effects are. As @felix91gr says, the same release might provide more confidence this will work, but I don't know any of the details of how the releases are configured (or any customizations in our CI), so I don't know how much of a concern that might be.

@millenomi If you do merge, can you keep an eye on the bots and revert ASAP if there are issues?

**Comment by millenomi** [comment]
We can retry but I have been unavailable to check on the patch's progress and CI for a little bit as other work priorities have weighted.

File: Foundation/FileManager.swift [hunk 1]
```diff
@@ -120,18 +122,333 @@ open class FileManager : NSObject {
...
     }
     
+    private enum _SearchPathDomain {
+        case system
+        case local
+        case network
+        case user
+        
+        init?(_ domainMask: SearchPathDomainMask) {
+            if domainMask == .systemDomainMask {
```
Comment (amosavian) [inline Foundation/FileManager.swift]: 1. domainMask is an OptionSet afaik, and urls function will return an array with multiple items if multiple masks are passes I think. I’m not sure however.

File: Foundation/FileManager.swift [hunk 2]
```diff
@@ -120,18 +122,333 @@ open class FileManager : NSObject {
...
+                self = .system; return
+            }
+            if domainMask == .localDomainMask {
```
Comment (amosavian) [inline Foundation/FileManager.swift]: Why not use switch case instead?

File: Foundation/FileManager.swift (hunk 1 above)
Comment (millenomi) [reply Foundation/FileManager.swift]: No, they're documented to accept a single mask at a time. Weirdness of the Darwin API, but yeah.

File: Foundation/FileManager.swift (hunk 2 above)
Comment (millenomi) [reply Foundation/FileManager.swift]: IIRC I tried, Swift complained, and I switched over without investigating.

File: Foundation/FileManager.swift (hunk 1 above)
Comment (millenomi) [reply Foundation/FileManager.swift]: My bad — you're correct, and this patch is wrong. (_url…_ accepts a single mask, but _urls…_ does not.)

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
Turns out it is indeed helpful if I remember to ask the bot for testing.

**Comment by millenomi** [comment]
This has gone on long enough. I'm going to hit that merge button, and we'll deal with bugs in future follow-ups.

**Pull request merged**